Content.Server/Traits/TraitSystem.Functions.cs

```diff
@@ -0,0 +1,280 @@
+        }
+    }
+}
+
+/// Used for traits that remove a component upon a player spawning in.
+[UsedImplicitly]
+public sealed partial class TraitRemoveComponent : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public ComponentRegistry Components { get; private set; } = new();
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        foreach (var (name, _) in Components)
+            entityManager.RemoveComponent(uid, (Component) factory.GetComponent(name));
+    }
+}
+
+/// Used for traits that add an action upon a player spawning in.
+[UsedImplicitly]
+public sealed partial class TraitAddActions : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public List<EntProtoId>? Actions { get; private set; } = default!;
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        if (Actions is null)
+            return;
+
+        var actionSystem = entityManager.System<SharedActionsSystem>();
+
+        foreach (var id in Actions)
+        {
+            EntityUid? actionId = null;
+            if (actionSystem.AddAction(uid, ref actionId, id))
+                actionSystem.StartUseDelay(actionId);
+        }
+    }
+}
+
+/// Used for traits that add an Implant upon spawning in.
+[UsedImplicitly]
+public sealed partial class TraitAddImplant : TraitFunction
+{
+    [DataField(customTypeSerializer: typeof(PrototypeIdHashSetSerializer<EntityPrototype>))]
+    [AlwaysPushInheritance]
+    public HashSet<string> Implants { get; private set; } = new();
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        var implantSystem = entityManager.System<SharedSubdermalImplantSystem>();
+        implantSystem.AddImplants(uid, Implants);
+    }
+}
+
+/// <summary>
+///     If a trait includes any Psionic Powers, this enters the powers into PsionicSystem to be initialized.
+///     If the lack of logic here seems startling, it's okay. All of the logic necessary for adding Psionics is handled by InitializePsionicPower.
+/// </summary>
+[UsedImplicitly]
+public sealed partial class TraitAddPsionics : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public List<string>? PsionicPowers { get; private set; } = default!;
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        var prototype = IoCManager.Resolve<IPrototypeManager>();
+        var psionic = entityManager.System<PsionicAbilitiesSystem>();
+        if (PsionicPowers is null)
+            return;
```

Comment:
```suggestion
```

Reply:
And even if you want to keep it nullable (I don't really understand why) you should move it to the very top. The guard clause should be before you get any data that is not used in the guard clause.